ConsoleApp/Program.cs

```diff
@@ -0,0 +1,443 @@
+                if (res > 0)
+                    l = m + 1;
+
+                // If x is smaller, ignore right half
+                else
+                    r = m - 1;
+            }
+
+            return -1;
+        }
+        private static void SearchByName()
+        {
+            Console.Clear();
+            Console.WriteLine("Search condition - if first/last name contains search term you will see it");
+
+            Console.WriteLine("Enter Search Term...");
+            var searchTerm = Console.ReadLine();
+
+            try
+            {
+                CheckEnteredParameter(ref searchTerm);
+
+                bool found = false;
+                foreach (var (firstName, lastName, phoneNumber) in ReadPhoneBook())
+                {
+                    if (firstName.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) ||
+                        lastName.Contains(searchTerm, StringComparison.OrdinalIgnoreCase))
+                    {
+                        Console.WriteLine($"Found user {firstName} {lastName} with phone {phoneNumber}");
+                        found = true;
+                    }
+                }
+
+                if (!found)
+                {
+                    throw new Exception("Not such users");
+                }
+            }
+            catch (Exception ex)
+            {
+                ShowExeption(ex.Message, "");
+            }
+            Wait();
+        }
+        private static int SearchByFullName(ref string searchFirstName, ref string searchLastName, ref (string, string, string)[] PhoneBook)
+        {
+
+            CheckEnteredParameters(ref searchFirstName, ref searchLastName);
+
+            int index = BinarySearchByLastName(PhoneBook, searchLastName);
+            if (index != -1)
+            {
+                if (searchFirstName == PhoneBook[index].Item2)
+                {
+                    Console.WriteLine($"Found user {searchFirstName} {searchLastName} with phone {PhoneBook[index].Item3}");
+                    index += 1;
+                }
+                else
+                {
+                    throw new Exception("The \"" + searchFirstName + " " + searchLastName + "\" user, not found!");
+                }
+            }
+
+            return index;
+        }
+        private static void SearchByPhone()
+        {
+            Console.Clear();
+            Console.WriteLine("Search condition - enter phone number");
+
+            Console.WriteLine("Enter Search Phone number...");
+            var searchPhone = Console.ReadLine();
+
+            try
+            {
+                CheckEnteredPhoneNumberParameter(searchPhone);
+
+                bool found = false;
+                if (searchPhone.Length > 1)
+                {
+                    foreach (var (firstName, lastName, phoneNumber) in ReadPhoneBook())
+                    {
+                        if (phoneNumber.Contains(searchPhone, StringComparison.OrdinalIgnoreCase))
```

Comment:
it's not critical, but you can do it without StringComparison.OrdinalIgnoreCase because it's phone number